File: 2024/02 February/Boolean Parenthesization/main.cpp
// Link: https://www.geeksforgeeks.org/problems/boolean-parenthesization5610/1

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int countWays(int n, string s)
    {

        vector<vector<int>> tru(n, vector<int>(n, 0));
        vector<vector<int>> fals(n, vector<int>(n, 0));

        int md = 1003;

        for (int itr = 0; itr < n; itr = itr + 2)
        {

            for (int i = 0; i + itr < n; i = i + 2)
            {

                int j = i + itr;

                if (i == j)
                {
                    tru[i][j] = s[i] == 'T' ? 1 : 0;
                    fals[i][j] = s[i] == 'F' ? 1 : 0;
                }

                for (int k = i + 1; k < j; k = k + 2)
                {

                    int left_tru = tru[i][k - 1];
                    int right_tru = tru[k + 1][j];
                    int left_fals = fals[i][k - 1];
                    int right_fals = fals[k + 1][j];

                    if (s[k] == '|')
                    {

                        tru[i][j] += left_tru * right_tru + left_tru * right_fals + right_tru * left_fals;
                        fals[i][j] += left_fals * right_fals;
                    }
                    else if (s[k] == '&')
                    {

                        tru[i][j] += left_tru * right_tru;
                        fals[i][j] += left_fals * right_fals + left_tru * right_fals + right_tru * left_fals;
                    }
                    else if (s[k] == '^')
                    {

                        tru[i][j] += left_tru * right_fals + right_tru * left_fals;
                        fals[i][j] += left_tru * right_tru + left_fals * right_fals;
                    }
                }

                tru[i][j] = (tru[i][j]) % md;
                fals[i][j] = (fals[i][j]) % md;
            }
        }

        return tru[0][n - 1];
    }
};

int main()
{
    Solution obj;
    int n = 7;
    string str = "T|T&F^T";
    cout << obj.countWays(n, str) << endl;
    n = 5;
    str = "T^F|F";
    cout << obj.countWays(n, str) << endl;
    return 0;
}

File: 2024/02 February/Boolean Parenthesization/main.py
# Link: https://www.geeksforgeeks.org/problems/boolean-parenthesization5610/1


class Solution:
    def countWays(self, n, s):
        dp = [[[0, 0] for i in range(n)] for j in range(n)]
        for i in range(0, n, 2):
            if s[i] == "T":
                dp[i][i][1] = 1
            else:
                dp[i][i][0] = 1
        for l in range(3, n + 1, 2):
            for i in range(0, n - l + 1, 2):
                j = i + l - 1
                t = (l - 1) // 2
                for k in range(1, t + 1):
                    k2 = 2 * k
                    if s[i + k2 - 1] == "&":
                        dp[i][j][1] += dp[i][i + k2 - 2][1] * dp[i + k2][j][1]
                        dp[i][j][0] += (
                            dp[i][i + k2 - 2][1] * dp[i + k2][j][0]
                            + dp[i][i + k2 - 2][0] * dp[i + k2][j][1]
                            + dp[i][i + k2 - 2][0] * dp[i + k2][j][0]
                        )
                    elif s[i + k2 - 1] == "|":
                        dp[i][j][0] += dp[i][i + k2 - 2][0] * dp[i + k2][j][0]
                        dp[i][j][1] += (
                            dp[i][i + k2 - 2][0] * dp[i + k2][j][1]
                            + dp[i][i + k2 - 2][1] * dp[i + k2][j][0]
                            + dp[i][i + k2 - 2][1] * dp[i + k2][j][1]
                        )
                    else:
                        dp[i][j][0] += (
                            dp[i][i + k2 - 2][0] * dp[i + k2][j][0]
                            + dp[i][i + k2 - 2][1] * dp[i + k2][j][1]
                        )
                        dp[i][j][1] += (
                            dp[i][i + k2 - 2][1] * dp[i + k2][j][0]
                            + dp[i][i + k2 - 2][0] * dp[i + k2][j][1]
                        )
        return dp[0][n - 1][1] % 1003


if __name__ == "__main__":
    obj = Solution()
    n = 7
    s = "T|T&F^T"
    print(obj.countWays(n, s))
    n = 5
    s = "T^F|F"
    print(obj.countWays(n, s))
